testify/test_rerunner.py

```diff
@@ -16,7 +19,16 @@ def __init__(self, *args, **kwargs):
         super(TestRerunner, self).__init__(*args, **kwargs)
 
     def discover(self):
-        for class_path, lines in groupby(self.rerun_test_file, lambda line: line.rpartition('.')[0]):
+        first_char = self.rerun_test_file.read(1)
```

Comment:
just read the entire file eagerly and do the splitlines here.  It simplifies your code significantly